test/switchboard/default-switchboards/FastSwitchboard.t.sol

```diff
@@ -5,104 +5,305 @@ import "../../Setup.t.sol";
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            0,
+            _watcherPrivateKey
+        );
+        assertTrue(fastSwitchboard.isAttested(_watcher, root));
+    }
 
-        assertTrue(fastSwitchboard.isAttested(watcher, packetId));
+    function testAttestInvalidRoot() external {
+        vm.expectRevert(InvalidRoot.selector);
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            1000, // Incorrect proposalCount
+            _watcherPrivateKey
+        );
     }
 
     function testDuplicateAttestation() external {
-        bytes32 digest = keccak256(
-            abi.encode(address(fastSwitchboard), _a.chainSlug, packetId)
+        vm.expectEmit(false, false, false, true);
+        emit ProposalAttested(packetId, 0, root, _watcher, 1);
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            0,
+            _watcherPrivateKey
+        );
+        assertTrue(fastSwitchboard.isAttested(_watcher, root));
+
+        vm.expectRevert(AlreadyAttested.selector);
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            0,
+            _watcherPrivateKey
+        );
+    }
+
+    function testDuplicateAttestationOnDuplicateProposal() external {
+        signAndPropose(_b.chainSlug, packetId, root);
+
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            0,
+            _watcherPrivateKey
+        );
+        vm.expectRevert(AlreadyAttested.selector);
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            1,
+            _watcherPrivateKey
+        );
+    }
+
+    function testAttestationOnDuplicateProposal() external {
+        signAndPropose(_b.chainSlug, packetId, root);
+
+        vm.expectEmit(false, false, false, true);
+        emit ProposalAttested(packetId, 0, root, _watcher, 1);
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            0,
+            _watcherPrivateKey
         );
-        bytes memory sig = _createSignature(digest, _watcherPrivateKey);
 
         vm.expectEmit(false, false, false, true);
-        emit PacketAttested(packetId, watcher);
+        emit ProposalAttested(packetId, 1, root, _altWatcher, 2);
+        _attestOnDst(
+            address(fastSwitchboard),
+            _b.chainSlug,
+            packetId,
+            1,
+            _altWatcherPrivateKey
+        );
 
-        fastSwitchboard.attest(packetId, sig);
+        bool isAllowed = fastSwitchboard.allowPacket(
+            root,
+            packetId,
+            0,
+            _a.chainSlug,
+            0
+        );
```

Comment:
allowPacket checks should be done with current timestamp. If propose time is passed as 0, packet becomes valid due to the delay even without enough attestations.